### PR TITLE
Add Support for Managing Service Level Objectives

### PR DESCRIPTION
**Pull request by platinummonkey**
Adds support for Service Level Objectives `/v1/slo` endpoints. 

Note: This does not include the status API currently, that is a work in flight and will come with a separate PR.

Also updates some packages from `datadog_test` to `datadog`. I'm not sure why these were different in the first place

### CODE REVIEW

File: service_level_objectives_test.go [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+		ts, c := testSLOGetMock(t2, "./tests/fixtures/service_level_objectives/delete_by_timeframe_response.json")
+		defer ts.Close()
+
+		resp, err := c.DeleteServiceLevelObjectiveTimeFrames(map[string][]string{
+			"12345678901234567890123456789012": {"7d"},
+			"abcdefabcdefabcdefabcdefabcdefab": {"30d", "90d"},
+		})
+		assert.NoError(t2, err)
+		assert.EqualValues(t2, resp.UpdatedIDs, []string{"12345678901234567890123456789012"})
+		assert.EqualValues(t2, resp.DeletedIDs, []string{"abcdefabcdefabcdefabcdefabcdefab"})
```
Comment (bkabrda) [inline service_level_objectives_test.go]: `abcdefabcdefabcdefabcdefabcdefab` has two timeframes: 7d and 30d. Why does this request delete it?

File: service_level_objectives.go [hunk 2]
```diff
@@ -0,0 +1,360 @@
...
+		uriValues.Set("offset", fmt.Sprintf("%d", offset))
+	}
+	// specify by query
+	if query != "" && len(ids) == 0 {
+		uriValues.Set("query", query)
+	}
+	// specify by `ids`
+	if query == "" && len(ids) > 0 {
+		uriValues.Set("ids", strings.Join(ids, ","))
+	}
```
Comment (bkabrda) [inline service_level_objectives.go]: So what should happen if `query != "" && len(ids) >0`? (I'm assuming that if `query == "" && len(ids) == 0`, then nothing needs to be done)?

File: integration/service_level_objectives_test.go [hunk 3]
```diff
@@ -0,0 +1,67 @@
+package integration
+
+import (
+	"testing"
+
+	"github.com/platinummonkey/go-datadog-api"
```
Comment (bkabrda) [inline integration/service_level_objectives_test.go]: You accidentally left your fork URL here.

File: integration/service_level_objectives_test.go [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+		},
+		Type: &datadog.ServiceLevelObjectiveTypeMetric,
+		Query: &datadog.ServiceLevelObjectiveMetricQuery{
+			Numerator:   datadog.String("sum:my.metric{type:good}.as_count()"),
+			Denominator: datadog.String("sum:my.metric{*}.as_count()"),
+		},
+	}
+
+	// Create
+	actual, err := client.CreateServiceLevelObjective(expected)
```
Comment (bkabrda) [inline integration/service_level_objectives_test.go]: Getting following error here:

```
integration/service_level_objectives_test.go:34:51: cannot use expected (type datadog.ServiceLevelObjective) as type *datadog.ServiceLevelObjective in argument to client.CreateServiceLevelObjective
```

File: service_level_objectives.go (hunk 2 above)
Comment (platinummonkey) [reply service_level_objectives.go]: Correct, both are optional

File: service_level_objectives.go [hunk 5]
```diff
@@ -0,0 +1,360 @@
...
+}
+
+type timeframesDeleteResp struct {
+	Data   *ServiceLevelObjectiveDeleteTimeFramesResponse `json:"data"`
+	Errors []*ServiceLevelObjectiveDeleteTimeFramesError  `json:"errors"`
+}
+
+// DeleteServiceLevelObjectiveTimeFrames will delete SLO timeframes individually.
+// This is useful if you have a SLO with 3 time windows and only need to delete some of the time windows.
+// It will do a full delete if all time windows are removed as a result.
```
[the remark this answers is not in the document]
Comment (platinummonkey) [reply service_level_objectives.go]: Full delete meaning if there are 0 time frames remaining then the entire SLO definition will be removed. Yes Update would also do partial deletes. The difference being this endpoint spans multiple SLOs powering the check boxes in the UI

File: service_level_objectives_test.go (hunk 1 above)
Comment (platinummonkey) [reply service_level_objectives_test.go]: In the end it will have 0 specified time frames.

Comment (bkabrda) [reply service_level_objectives_test.go]: Then I probably don't understand what the argument passed to `DeleteServiceLevelObjectiveTimeFrames` means. Intuitively, I'd expect it to delete `30d` and `90d` for SLO `abcdefabcdefabcdefabcdefabcdefab` and `7d` for SLO `12345678901234567890123456789012`. Does it do something different?

File: service_level_objectives.go (hunk 2 above)
Comment (platinummonkey) [reply service_level_objectives.go]: fixing this inline docs and better organization

File: service_level_objectives_test.go (hunk 1 above)
Comment (platinummonkey) [reply service_level_objectives_test.go]: it does do that exactly that. I'll add more comments here, but the idea is that: `12345678901234567890123456789012` has 2 (or more time frames) already specified, and `abcdefabcdefabcdefabcdefabcdefab` has *only* those 2 timeframes specified.

so in the case of `abcdefabcdefabcdefabcdefabcdefab` after deleting the 2 specified there will be 0 timeframes remaining, hence it's fully deleted

in the case of `12345678901234567890123456789012` it still has 1 timeframe remaining

File: integration/service_level_objectives_test.go (hunk 4 above)
Comment (platinummonkey) [reply integration/service_level_objectives_test.go]: should be fixed, see change above

File: service_level_objectives.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (platinummonkey) [reply service_level_objectives.go]: added more docs with example here and in the test

File: service_level_objectives.go [hunk 6]
```diff
@@ -0,0 +1,360 @@
...
+
+	if err := client.doJsonRequest("PUT", fmt.Sprintf("/v1/slo/%s", slo.GetID()), slo, &out); err != nil {
+		return err
+	}
+
+	if out.Error != "" {
+		return fmt.Errorf(out.Error)
+	}
+
+	if len(out.Data) != 1 {
```
[the remark this answers is not in the document]
Comment (platinummonkey) [reply service_level_objectives.go]: removed here and below

File: service_level_objectives_test.go (hunk 1 above)
Comment (bkabrda) [reply service_level_objectives_test.go]: Well that's my point exactly. If you look at the fixture for `abcdefabcdefabcdefabcdefabcdefab`, you're going to see that it has `7d` and `30d`, so by the command above you'd only delete its `30d` timeframe, but leave the `7d` one. You probably meant to have `"abcdefabcdefabcdefabcdefabcdefab": {"7d", "30d"}` in the request.

**Comment by platinummonkey** [comment]
Context for schemas at the simplest validation level (jsonschema):

Create Schema: 
```json
{
        "type": "object",
        "properties": {
            "type": {"type": "string", "enum": ["metric", "monitor"]},
            "name": {"type": "string", "minLength": 1, "maxLength": 16384},
            "monitor_ids": {
                "type": "array",
                "maxItems": 100,
                "uniqueItems": true,
                "items": {"type": "number", "exclusiveMinimum": 0},
            },
            "monitor_search": {"type": "string", "minLength": 1, "maxLength": 16384},
            "query": {
                "type": "object",
                "properties": {
                    "numerator": {"type": "string", "minLength": 1, "maxLength": 65536},
                    "denominator": {"type": "string", "minLength": 1, "maxLength": 65536},
                },
            },
            "tags": {
                "type": "array",
                "maxItems": 128,
                "uniqueItems": true,
                "items": {"type": "string", "minLength": 1, "maxLength": 256},
            },
            "thresholds": {
                "type": "object",
                "patternProperties": {
                    "^[0-9]+[a-z]+$": {
                        "type": "object",
                        "properties": {
                            "slo": {"type": "number", "minimum": 0, "maximum": 100},
                            "warning": {"type": "number", "minimum": 0, "maximum": 100},
                        },
                        "required": ["slo"],
                    }
                },
            },
        },
        "required": ["type", "name", "thresholds"],
    }
```

Update Schema:
```json
{
        "type": "object",
        "properties": {
            "type": {"type": "string", "enum": ["metric", "monitor"]},
            "name": {"type": "string", "minLength": 1, "maxLength": 16384},
            "monitor_ids": {
                "type": "array",
                "maxItems": 100,
                "uniqueItems": true,
                "items": {"type": "number", "exclusiveMinimum": 0},
            },
            "monitor_search": {"type": "string", "minLength": 1, "maxLength": 16384},
            "query": {
                "type": "object",
                "properties": {
                    "numerator": {"type": "string", "minLength": 1, "maxLength": 65536},
                    "denominator": {"type": "string", "minLength": 1, "maxLength": 65536},
                },
            },
            "tags": {
                "type": "array",
                "maxItems": 128,
                "uniqueItems": true,
                "items": {"type": "string", "minLength": 1, "maxLength": 256},
            },
            "thresholds": {
                "type": "object",
                "patternProperties": {
                    "^[0-9]+[a-z]+$": {
                        "type": "object",
                        "properties": {
                            "slo": {"type": "number", "minimum": 0, "maximum": 100},
                            "warning": {"type": "number", "minimum": 0, "maximum": 100},
                        },
                        "required": ["slo"],
                    }
                },
            },
        },
        "required": ["type"],
    }
```

ids_schema:
```json
{
        "type": "array",
        "minItems": 1,
        "maxItems": 1000,
        "uniqueItems": True,
        "items": {"type": "string", "minLength": 32, "maxLength": 32},
    }
```

File: service_level_objectives_test.go (hunk 1 above)
Comment (platinummonkey) [reply service_level_objectives_test.go]: I'm not sure what fixture you're referring to here. The fixtures are only examples of the responses not the states prior. I've updated the comments in this test to relay the state prior to the test method.

Comment (bkabrda) [reply service_level_objectives_test.go]: You created a fixture `tests/fixtures/service_level_objectives/create_response_metric.json` which contains an SLO with name `abcdefabcdefabcdefabcdefabcdefab`. This made it really confusing, as I thought this test case was also related to it and that didn't make sense. The added comment clears this up, so I have no further objections here.

**Pull request merged**